### ISSUE TITLE
sql: session-based tracing performance regression

### ISSUE THREAD

**Issue by petermattis** [cockroachdb/cockroach#16518]
```
~/Development/go/src/github.com/cockroachdb/cockroach/pkg/sql (d290370...) benchstat out.good out.bad
name            old time/op    new time/op    delta
KV/Scan1_SQL-8     188µs ± 2%     225µs ± 0%  +20.04%  (p=0.000 n=10+9)

name            old alloc/op   new alloc/op   delta
KV/Scan1_SQL-8    28.8kB ± 0%    35.8kB ± 0%  +23.94%  (p=0.000 n=10+8)

name            old allocs/op  new allocs/op  delta
KV/Scan1_SQL-8       268 ± 0%       402 ± 0%  +50.37%  (p=0.000 n=10+10)
```

Bad is 6a369c3f40448b782cd6b047cd7915ac907a2087 (#16253). Good is d290370ac795f3efd1cc5036a003505e2f2eb43d.  I haven't investigated what is going on yet. Notice the massive increase in the number of allocations. Session-based tracing is off by default, right?

@tamird Did the new benchmark storage infrastructure catch this? Even if it didn't, this is a good example on which we should backfill some data to verify we would notice a performance regression.

**Comment by petermattis** [comment]
This diff "fixes" the regression:

```diff
~/Development/go/src/github.com/cockroachdb/cockroach/pkg master git diff
diff --git a/pkg/sql/session.go b/pkg/sql/session.go
index 1972e5b..8b1b504 100644
--- a/pkg/sql/session.go
+++ b/pkg/sql/session.go
@@ -860,7 +860,8 @@ func (ts *txnState) resetForNewSQLTxn(e *Executor, s *Session, implicitTxn bool)
                        opName, opentracing.ChildOf(parentSp.Context()), tracing.Recordable)
        } else {
                // Create a root span for this SQL txn.
-               sp = tracer.StartSpan(opName, tracing.Recordable)
+               // sp = tracer.StartSpan(opName, tracing.Recordable)
+               sp = tracer.StartSpan(opName)
        }

        // Start recording for the traceTxnThreshold and debugTrace7881Enabled
@@ -882,9 +883,9 @@ func (ts *txnState) resetForNewSQLTxn(e *Executor, s *Session, implicitTxn bool)
        // Put the new span in the context.
        ctx = opentracing.ContextWithSpan(ctx, sp)

-       if !tracing.IsRecordable(sp) {
-               log.Fatalf(ctx, "non-recordable transaction span of type: %T", sp)
-       }
+       // if !tracing.IsRecordable(sp) {
+       //      log.Fatalf(ctx, "non-recordable transaction span of type: %T", sp)
+       // }

        ts.sp = sp
        ts.Ctx = ctx
```

So something about always creating a Recordable span is problematic. I'm not familiar enough with this new code to know why we're creating a Recordable span even when tracing is disabled.

**Comment by andreimatei** [comment]
cc @RaduBerinde 

Hmm, looking. The reason why we start a `Recordable` span even when tracing is off is because tracing might become on later on in that transaction, and we want to be able to record stuff happening after the point where tracing becomes on. A `Recordable` span has some allocation cost, but it's not supposed to be big.
I hope what's going on here is that there's an unexpected inefficiency, and not that this benchmark is micro-enough that a small number of allocations throw it off so much :S.
Looking.

**Comment by RaduBerinde** [comment]
Could we use a non-recordable span by default, and when tracing is enabled close that span and open a new recordable one?

Either way, the overhead should be investigated, it shouldn't be that much.. It should be one span allocation per request, plus an extra allocation (`sso`) for each `StartSpan` call.

**Comment by andreimatei** [comment]
We can account for the majority of new allocations in an interaction between logging and tracing that's easy to fix, and some others in the creation of the spans that we can probably also eliminate.
There might be a few left that are unclear, so still looking.

**Comment by petermattis** [comment]
> The reason why we start a Recordable span even when tracing is off is because tracing might become on later on in that transaction, and we want to be able to record stuff happening after the point where tracing becomes on.

While this is interesting functionality, I wouldn't mind if we temporarily disabled it and only allow tracing to be enabled at the start of or outside a transaction. Or take @RaduBerinde's suggestion and replace the non-recordable span with a recordable one when tracing is enabled.

Don't let this comment get in the way of you fixing the underlying problem, though.

**Comment by andreimatei** [comment]
I've looked with Radu a bunch. It looks like we can get rid of ~all of the new allocations due to session tracing, taking us from ~400 allocs/op to ~300 allocs/op. This is through a combination of being smarter about logging in `Recordable`, but not "recording", spans, and general optimizations to the tracing infrastructure. Although not particularly important for this benchmark, the latter tracing optimizations will surely pay off in the future. I'll send these out.

Now, the difference between ~300 and the original ~270 is not my PR :). Peter has compared  6a369c3 (#16253) - bad - with d290370 - good -. d290370 is a commit by Radu that has been merged just before I merged my PR, but the commit itself is quite old - 12 days or so. There were other allocation regressions in there, at least two.
I've bisected to see what's going on and I found the bigger of the two and filed #16531.
